src/player.cpp

```diff
@@ -6868,72 +6835,16 @@ void player::use_fire(const int quantity)
 
     if( g->m.has_nearby_fire( pos() ) ) {
         return;
-    } else if (has_charges("torch_lit", 1)) {
-        return;
-    } else if (has_charges("battletorch_lit", 1)) {
-        return;
-    } else if (has_charges("handflare_lit", 1)) {
-        return;
-    } else if (has_charges("candle_lit", 1)) {
-        return;
-    } else if (has_charges("shishkebab_on", quantity)) {
-        return;
-    } else if (has_charges("firemachete_on", quantity)) {
-        return;
-    } else if (has_charges("broadfire_on", quantity)) {
-        return;
-    } else if (has_charges("firekatana_on", quantity)) {
-        return;
-    } else if (has_charges("zweifire_on", quantity)) {
-        return;
-    } else if (has_charges("ref_lighter", quantity)) {
-        use_charges("ref_lighter", quantity);
-        return;
-    } else if (has_charges("matches", quantity)) {
-        use_charges("matches", quantity);
-        return;
-    } else if (has_charges("lighter", quantity)) {
-        use_charges("lighter", quantity);
-        return;
-    } else if (has_charges("flamethrower", quantity)) {
-        use_charges("flamethrower", quantity);
-        return;
-    } else if (has_charges("flamethrower_simple", quantity)) {
-        use_charges("flamethrower_simple", quantity);
-        return;
-    } else if (has_charges("hotplate", quantity)) {
-        use_charges("hotplate", quantity);
-        return;
-    } else if (has_charges("welder", quantity)) {
-        use_charges("welder", quantity);
-        return;
-    } else if (has_charges("welder_crude", quantity)) {
-        use_charges("welder_crude", quantity);
-        return;
-    } else if (has_charges("shishkebab_off", quantity)) {
-        use_charges("shishkebab_off", quantity);
-        return;
-    } else if (has_charges("firemachete_off", quantity)) {
-        use_charges("firemachete_off", quantity);
-        return;
-    } else if (has_charges("broadfire_off", quantity)) {
-        use_charges("broadfire_off", quantity);
-        return;
-    } else if (has_charges("firekatana_off", quantity)) {
-        use_charges("firekatana_off", quantity);
-        return;
-    } else if (has_charges("zweifire_off", quantity)) {
-        use_charges("zweifire_off", quantity);
-        return;
-    } else if (has_active_bionic( bio_tools ) && power_level > quantity * 5 ) {
-        charge_power( -quantity * 5 );
-        return;
-    } else if (has_bionic( bio_lighter ) && power_level > quantity * 5 ) {
-        charge_power( -quantity * 5 );
-        return;
-    } else if (has_bionic( bio_laser ) && power_level > quantity * 5 ) {
-        charge_power( -quantity * 5 );
+    } else if( has_item_with_flag( "FIRE" ) ) {
         return;
+    } else if( has_item_with_flag( "FIRESTARTER" ) ) {
+        auto firestarters = all_items_with_flag( "FIRESTARTER" );
+        for( auto &i : firestarters ) {
+            if( has_charges( i->typeId() , quantity ) ) {
```

Comment:
The copy-pasted snippet can be eliminated by introducing a function which accepts `quantity` and returns a suitable firestarter (or the null item if none was found). Also, `std::any()` would be cleaner for this purpose.

Reply:
There are two functions `has_fire` and `use_fire` which are only called directly by `has_charges` and `use_charges` functions.

![image](https://user-images.githubusercontent.com/16213433/35282136-ccb115f4-0065-11e8-8604-dfad4ee2622e.png)

Do you suggest combining these two functions to a single one `bool player::use_or_has_fire( const int quantity, const bool )`  to get rid of duplicate code and make existing functions just wrappers of a new one?

What's with `std::any()`? I'm not familiar with it. Is it some construction which can return multiple types from same function?

Reply:
> Do you suggest combining these two functions to a single one bool 

Not exactly, I had something like this in mind:

```c++
itype player::find_firestarter( int quantity ) const
{
    const auto firestarters = all_items_with_flag( "FIRESTARTER" );
    const auto it = std::find_if( firestarters.begin(), firestarters.end(), [quantity]( const item *elem ) {
	return elem->has_charges( quantity );
    } ); 
    return it == firestarters.end() ? ret_null : it->typeId();
}

    ...

    const itype firestarter = find_firestarter( quantity );
    if( firestarter != "null" ) {	// should become a proper id eventually.
        use_charges( firestarter, quantity );
    }
    ...
```

> What's with std::any()? I'm not familiar with it. Is it some construction which can return multiple types from same function?

It returns iterators and generally useful as a replacement for `for` loops followed by conditions and `return` statements.

Reply:
> It returns iterators and generally useful as a replacement for for loops followed by conditions and return statements.

So, is it this - http://www.cplusplus.com/reference/algorithm/any_of/ ?

Reply:
> ````itype player::find_firestarter( int quantity ) const````

I guess we should return `item`, not `itype`?

Reply:
> So, is it this

Yes, it is. I mistyped the name, so thanks.

> I guess we should return item, not itype?

No, we probably shouldn't. The whole point of passing `itype` is that more than one instance of `item` can be used to acquire the needed amount of charges.

Reply:
> No, we probably shouldn't. The whole point of passing itype is that more than one instance of item can be used to acquire the needed amount of charges.

So `item` is certain instance of some item and `itype` will return item  of some type (possibly even multiple items)?